sycl/doc/design/OffloadDesign.md

```diff
@@ -0,0 +1,396 @@
+# Implementation design for offloading model
+
+## Introduction
+
+This document covers the implementation design for using the new offloading
+model for the DPC++ Compiler.  This leverages the existing community Offloading 
+design [OffloadingDesign][1] which covers the Clang driver and code generation
+steps for creating offloading applications.
+
+[1]: <../../../clang/docs/OffloadingDesign.rst>
+
+The current offloading model is completely encapsulated within the Clang
+Compiler Driver requiring the driver to perform all of the additional steps
+for generating the host and device compilation passes.  The Driver is also
+responsible for performing any of the link-time processing that occurs for
+each device target.
+
+The updated offloading model removes much of the functionality that is
+performed during the link phase of the offloading compilation and moves it
+to a `clang-linker-wrapper` tool.
+
+Below is a general representation of the overall offloading flow that is
+performed during a full compilation from source to final executable.  The
+compiler driver is responsible for creating the fat object and the
+`clang-linker-wrapper` tool is responsible for the general functionality that is
+performed during the link.
+
+![High level view of the offloading flow](images/OffloadGeneralFlow.svg
+ "General Offload Flow")
+
+*Diagram 1: General Offload Flow*
+
+More specific details of the clang-linker-wrapper behavior can be seen in the
+[clang-linker-wrapper](#clang-linker-wrapper) section.
+
+## Fat Binary Representation
+
+Binaries generated during the offload compilation will be 'bundled' together
+to create a conglomerate fat binary.  Depending on the type of binary, the
+packaging of the device section will be different.
+
+## Device only Binaries
+
+Generation and consumption of device only binaries is also available.  The
+device only binaries are represented by the file generated from the
+[Packager](#packager) to generate an 'Offload Binary'.
+
+## Fat Binary Generation
+
+The generation of the fat binary will take place in the driver.  The model from
+the community is generating the fat binary as a secondary process when creating
+the host object.  Generation will be separated from the host compilation step.
+This is being done to enable proper support for using an external host compiler
+as well as taking advantage of potential parallelism during compilation of both
+the host and target device binaries.
+
+The fat object in the new model is generated during the host compilation.
+The host compilation takes an additional argument which points to the device
+binary which will be embedded in the final object.  Generation will be
+separated out to allow for potential parallelism during compilation of both the
+host and target device binaries.
+
+![Creating the fat object](images/OffloadFatObject.svg
+ "Fat Object Generation")
+
+*Diagram 2: Fat Object Generation*
```

Comment:
In the diagram, I think there are two changes needed: (1) Show host compilation using a separate box and (2) Output of clang-offload-packager should go into the 'link' box. and the output of the link box.
This is based on my understanding. I may have missed something here.

Thanks

Reply:
I have updated the diagram to be a little clearer.  Not sure why there was a 'link' section as there is no linking when generating fat objects.